ui/src/criteria/concept.tsx

```diff
@@ -171,48 +201,67 @@ function ConceptEdit(props: ConceptEditProps) {
 
             // Copy over existing children in case they're being loaded in
             // parallel.
+            let childChildren = data[id]?.children;
+            if (!childChildren) {
+              if (!hierarchy) {
+                hasChildren = !!entity.listChildren && !listChildrenEntity;
+              }
+              childChildren = !hasChildren ? [] : undefined;
+            }
+
             data[id] = {
               data: row,
-              children: data[id]?.children || (!hasChildren ? [] : undefined),
+              children: childChildren,
             };
           });
         }
 
-        if (id) {
+        if (parentId) {
           // Store children even if the data isn't loaded yet.
-          data[id] = { ...data[id], children };
+          data[parentId] = { ...data[parentId], children };
         } else {
-          data.root = { children, data: {} };
+          data.root = {
+            children: [...(data?.root?.children || []), ...children],
+            data: {
+              [ENTITY_ATTRIBUTE]: entity.name,
+            },
+          };
         }
       });
     },
     []
   );
 
-  const conceptsState = useAsyncWithApi<void>(
-    useCallback(() => {
-      updateData(() => ({}));
-      return api
-        .searchEntityInstances(
-          searchFilter(
-            props.data,
-            props.cohort.underlayName,
-            !hierarchyPath ? query : ""
+  const fetch = useCallback(() => {
```

Comment:
nit: I might advise against `fetch` as a local variable name, since it's shadowing the standard browser global `fetch`

Reply:
Done, good call.